packages/block-library/src/image/index.php

```diff
@@ -0,0 +1,357 @@
+ */
+function gutenberg_get_image_width_height( $block_attributes, $image_meta = null ) {
+	$block_witdh = gutenberg_get_block_width();
+
+	if ( empty( $block_attributes['fileWidth'] ) || empty( $block_attributes['fileHeight'] ) ) {
+		if ( ! empty( $block_attributes['width'] ) && ! empty( $block_attributes['height'] ) ) {
+			// Old post with an image that was resized?
+			return array(
+				$block_attributes['width'],
+				$block_attributes['height'],
+			);
+		} else {
+			// Image data is missing.
+			return false;
+		}
+	} else {
+		$image_file_width = (int) $block_attributes['fileWidth'];
+		$image_file_height = (int) $block_attributes['fileHeight'];
+	}
+
+	if ( ! empty( $image_meta ) &&
+		$image_file_width < $image_meta['width'] &&
+		! empty( $image_meta['sizes']['large'] ) &&
+		$image_file_width === (int) $image_meta['sizes']['large']['width'] &&
+		wp_image_matches_ratio( $image_file_width, $image_file_height, $image_meta['width'], $image_meta['height'] )
+	) {
+		// TODO: To reduce (fix) the need for upscaling or using the "full" size
+		// add `xlarge` image size generated by default!!
+		// Photos taken with modern phones and cameras are usually several megabytes in size, can be over 8MB in some cases.
+		// That makes the "full" image size not suitable for web use.
+		// Do not constrain images with `wide` and `full` allignment to the "large" image size.
+		// Sometimes need to use a larger image there, if one exists.
+		// Note: there is a constrain on the largest image size in the srcset attribute, see `wp_calculate_image_srcset()`.
+		$full_file_width = (int) $image_meta['width'];
+	} else {
+		$full_file_width = $image_file_width;
+	}
+
+	if ( $block_attributes['align'] === 'full' ) {
+		$image_width = min( (int) $block_witdh['full'], $full_file_width );
+	} elseif ( $block_attributes['align'] === 'wide' ) {
+		$image_width = min( (int) $block_witdh['wide'], $full_file_width );
+	} else {
+		$image_width = (int) $block_witdh['default'];
+
+		if ( ! empty( $block_attributes['width'] ) && ! empty( $block_attributes['editWidth'] ) ) {
+			// Scale the image if it was resized in the editor.
+			$scale = $block_attributes['width'] / $block_attributes['editWidth'];
+			// TODO: allow upscaling? Looks ugly, but someone may expect it?
+			$image_width = round( $image_width * $scale );
+		}
+
+		$image_width = min( $image_width, $image_file_width );
+	}
+
+	// Recalculate.
+	$image_size = wp_constrain_dimensions( $image_file_width, $image_file_height, $image_width );
+
+	/**
+	 * Filters the calculated image size for the image block.
+	 *
+	 * @since 4.2.0
+	 *
+	 * @param array $image_size The calculated image size width and height (in that order).
+	 * @param array $block_attributes The block attributes.
+	 * @param array $image_meta The image attachment meta data.
+	 */
+	return apply_filters( 'gutenberg_get_image_width_height', $image_size, $block_attributes, $image_meta );
```

Comment:
As a matter of practice, are we using `gutenberg_` as a prefix for filters in core, or are using something more generic, like `block_editor_`?